core/src/main/java/org/pentaho/di/compatibility/Value.java

```diff
@@ -0,0 +1,3960 @@
+          } else {
+            String string = getBigNumber().toString();
+            dos.writeInt( string.length() );
+            dos.writeChars( string );
+          }
+          break;
+        case VALUE_TYPE_DATE:
+          dos.writeBoolean( getDate() != null );
+          if ( getDate() != null ) {
+            dos.writeLong( getDate().getTime() );
+          }
+          break;
+        case VALUE_TYPE_NUMBER:
+          dos.writeDouble( getNumber() );
+          break;
+        case VALUE_TYPE_BOOLEAN:
+          dos.writeBoolean( getBoolean() );
+          break;
+        case VALUE_TYPE_INTEGER:
+          dos.writeLong( getInteger() );
+          break;
+        default:
+          break; // nothing
+      }
+    }
+  }
+
+  /**
+   * Write the value, including the meta-data to a DataOutputStream
+   *
+   * @param outputStream
+   *          the OutputStream to write to .
+   * @throws KettleFileException
+   *           if something goes wrong.
+   */
+  public void write( OutputStream outputStream ) throws KettleFileException {
+    try {
+      writeObj( new DataOutputStream( outputStream ) );
+    } catch ( Exception e ) {
+      throw new KettleFileException( "Unable to write value to output stream", e );
+    }
+  }
+
+  /**
+   * Read the metadata and data for this Value object from the specified data input stream
+   *
+   * @param dis
+   * @throws IOException
+   */
+  public void readObj( DataInputStream dis ) throws IOException {
+    // type
+    int theType = dis.readInt();
+    newValue( theType );
+
+    // name-length
+    int nameLength = dis.readInt();
+
+    // name
+    StringBuilder nameBuffer = new StringBuilder();
+    for ( int i = 0; i < nameLength; i++ ) {
+      nameBuffer.append( dis.readChar() );
+    }
+    setName( new String( nameBuffer ) );
+
+    // length & precision
+    setLength( dis.readInt(), dis.readInt() );
+
+    // Null?
+    setNull( dis.readBoolean() );
+
+    // Read the values
+    if ( !isNull() ) {
+      switch ( getType() ) {
+        case VALUE_TYPE_STRING:
+          // read the length
+          int stringLength = dis.readInt();
+          if ( stringLength < 0 ) {
+            setValue( (String) null );
+          } else {
+            byte[] chars = new byte[stringLength];
+            dis.readFully( chars );
+            setValue( new String( chars, Const.XML_ENCODING ) );
+          }
+          break;
+        case VALUE_TYPE_BIGNUMBER:
+          // read the length
+          int bnLength = dis.readInt();
+          if ( bnLength < 0 ) {
+            setValue( (BigDecimal) null );
+          } else {
+            StringBuilder buffer = new StringBuilder();
+            for ( int i = 0; i < bnLength; i++ ) {
+              buffer.append( dis.readChar() );
+            }
+            setValue( buffer.toString() );
+            try {
+              convertString( VALUE_TYPE_BIGNUMBER );
+            } catch ( KettleValueException e ) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)